pkg/pillar/cmd/README.md

```diff
@@ -0,0 +1,273 @@
+# Agent Services
+
+This directory contains agent services for pillar.
+
+Each subdirectory contains exactly one agent, whose name is the name of the directory.
+
+Look at [verifier](./verifier/) for an example of the implementation.
+
+## Design Pattern
+
+The design pattern for each agent is as follows. When creating a new agent, ensure
+that you follow the design pattern.
+
+The package name in the agent's directory is the same as the directory name.
+It should contain at least the following files:
+
+* `README.md` describing the agent and its purpose.
+* `run.go` containing the main library entry point for the agent.
+
+In addition, it should contain the following subdirectories:
+
+* `lib/` contains core logic, has no dependency on being called command-line or pubsub, unless pubsub is part of the core business logic. It is important to keep these functions as simple and straightforward as possible. Unit tests of business logic should be here as well.
+* `pubsub/` contains functions to handle pubsub messages, including all handlers. There should be no business logic here, only pubsub message handling. For business logic, handlers should call business logic functions in `lib/`.
+* `cmd/` contains CLI commands and sub-commands, all in package `main`. Can be called as `go run ./cmd`. Call either functions in `pubsub/` or `lib/` depending on the command.
+* `run.go` is the entrypoint for calling as part of zedbox, exposes the implementation of [`types.AgentRunner`](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar/types#AgentRunner).
+
+To run as a CLI:
+
+```bash
+go run ./cmd
+```
+
+To include as a pubsub library elsewhere, like in pillar, for example `verifier`:
+
+```go
+import (
+    "github.com/lf-edge/eden/pkg/verifier"
+)
+
+func main() {
+    verifier.Run(pubSubImpl, logrusLogger, baseLogObject, []string{argument1, argument2})
+}
+```
+
+## Agent entry point
+
+The agent entry point is `run.go`. This file should contain the following:
+
+```go
+package verifier
+
+import (
+    "github.com/lf-edge/eve/pkg/pillar/base"
+    myagent "github.com/lf-edge/eve/pkg/pillar/cmd/myagent/pubsub"
+    "github.com/lf-edge/eve/pkg/pillar/pubsub"
+    "github.com/sirupsen/logrus"
+)
+
+func Run(ps *pubsub.PubSub, logger *logrus.Logger, log *base.LogObject, arguments []string, baseDir string) int {
+    ctx := myagentpubsub.NewContext(ps, logger, log)
+    return ctx.Run(arguments, baseDir)
+}
+```
+
+where `myagent` should be replaced with the name of the agent.
+
+The `myagent/pubsub/` package should export a `NewContext()` function, which
+should return an agent-specific `context` object, which exports
+a `Run()` function that starts the agent.
+
+The `context.Run()` function contains all of the initialization and
+pubsub interaction.
+
+## CLI functions
```

Comment:
We should probably create some dev container for running these CLIs because most microservices have some dependencies on installed software (dhcpcd, dnsmasq, etc.) and could make some config changes and break something in the host.

Reply:
Those are exactly the dependencies I want to see. But not all of them have it. So if you are working with downloader and need only an http server or registry for testing, you should be able to test with just those, either listed or (even better, as you recommended) a devcontainer. If you are working with NIM, you might need dhcpcd and dnsmasq, etc.

As we get to those with greater complexities, we will start to draw those out. I really like the devcontainer approach.